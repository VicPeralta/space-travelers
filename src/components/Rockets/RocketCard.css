.rocket-card {
  display: flex;
  gap: 20px;
}

.rocket-card img {
  width: 200px;
}

.info-rocket {
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.rocket-card button {
  background-color: var(--button-rockets-background, dodgerblue);
  color: var(--button-rockets-color, white);
  width: 150px;
  padding: 10px 3px;
  border-radius: 8px;
  font-family: Arial, Helvetica, sans-serif;
  cursor: pointer;
}

.rocket-reserved {
  background-color: var(--badge-rockets-background);
  color: var(--badge-rockets-color);
  margin-right: 10px;
  padding-inline: 8px;
  border-radius: 5px;
  font-size: 0.8rem;
}

.rocket-card button:hover {
  font-size: 1.03rem;
  width: 150px;
}

@media (max-width: 390px) {
  .rocket-card {
    flex-direction: column;
  }

  .rocket-card img {
    width: auto;
  }
}
