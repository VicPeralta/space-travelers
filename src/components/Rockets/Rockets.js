import { useSelector } from 'react-redux';
import RocketList from './RocketsList';

const Rockets = () => {
  const rocketsList = useSelector((state) => state.rocketsReducer.rockets);
  const fetching = useSelector((state) => state.rocketsReducer.fetching);
  return (
    <>
      {!fetching && (
        <RocketList rockets={rocketsList} />
      )}
      {fetching && (
        <h3>Fetching data please wait...</h3>
      )}
    </>
  );
};

export default Rockets;
