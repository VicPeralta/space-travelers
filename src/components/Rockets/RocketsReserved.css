.reserved-container {
  min-width: 40%;
  margin: 20px;
}

.rockets-reserved {
  list-style: none;
  margin: 10px 0;
  padding: 0;
  border: 1px solid lightgray;
  border-radius: 8px;
}

.reserved-container li {
  margin: 0;
  padding: 10px 40px;
  border-bottom: 1px solid lightgray;
}

.reserved-container li:last-child {
  border: none;
}

@media (max-width: 390px) {
  .reserved-container {
    width: 340px;
    min-width: 300px;
    margin-inline: auto;
  }

  .reserved-container li {
    padding: 10px;
  }
}
