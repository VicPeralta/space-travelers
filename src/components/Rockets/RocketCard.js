import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { reserveRocket, cancelRocket } from '../../redux/rockets/rockets';
import './RocketCard.css';

const RocketCard = ({
  id, name, description, image, reserved,
}) => {
  const dispatch = useDispatch();
  const onClickReserved = (event) => {
    dispatch(reserveRocket(event.target.dataset.id));
  };
  const onClickCancel = (event) => {
    dispatch(cancelRocket(event.target.dataset.id));
  };
  return (
    <div className="rocket-card">
      <img src={image} alt="rocket type" width={200} height={150} />
      <div className="info-rocket">
        <h3>
          {name}
        </h3>
        <p>
          {reserved && (<span className="rocket-reserved">Reserved</span>)}
          {description}
        </p>
        {!reserved && (<button type="button" data-id={id} onClick={onClickReserved}>Reserve Rocket</button>)}
        {reserved && (<button type="button" data-id={id} onClick={onClickCancel}>Cancel Reservation</button>)}
      </div>
    </div>
  );
};

RocketCard.defaultProps = {
  reserved: false,
};

RocketCard.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  reserved: PropTypes.bool,
};
export default RocketCard;
