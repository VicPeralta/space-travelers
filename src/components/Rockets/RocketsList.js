import PropTypes from 'prop-types';
import RocketCard from './RocketCard';
import './RocketsList.css';

const RocketList = ({ rockets }) => (
  <div className="rocketContainer">
    {
      rockets.map((rocket) => (
        <RocketCard
          key={rocket.id}
          id={rocket.id}
          name={rocket.name}
          description={rocket.description}
          reserved={rocket.reserved}
          image={rocket.image}
        />
      ))
    }
  </div>
);

RocketList.propTypes = {
  rockets: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    description: PropTypes.string,
    image: PropTypes.string,
    reserved: PropTypes.bool,
  })).isRequired,
};
export default RocketList;
