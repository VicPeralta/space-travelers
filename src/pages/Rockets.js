import Rockets from '../components/Rockets/Rockets';

const RocketsPage = () => (
  <>
    <h2>Rockets</h2>
    <Rockets />
  </>
);

export default RocketsPage;
