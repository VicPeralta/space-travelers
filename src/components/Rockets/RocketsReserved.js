import PropTypes from 'prop-types';
import './RocketsReserved.css';

const RocketsReserved = ({ rockets }) => (
  <div className="reserved-container">
    <h2>My Rockets</h2>
    {rockets.length === 0 && (
      <h3> There are no rockets reserved</h3>
    )}
    {rockets.length > 0 && (
      <ul className="rockets-reserved">
        {
          rockets.map((rocket) => (
            <li key={rocket.id}>{rocket.name}</li>
          ))
        }
      </ul>
    )}
  </div>
);

RocketsReserved.propTypes = {
  rockets: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    description: PropTypes.string,
    image: PropTypes.string,
    reserved: PropTypes.bool,
  })).isRequired,
};
export default RocketsReserved;
