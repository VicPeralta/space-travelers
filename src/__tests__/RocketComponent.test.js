import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Rockets from '../components/Rockets/Rockets';
import RocketCard from '../components/Rockets/RocketCard';
import RocketList from '../components/Rockets/RocketsList';
import RocketsReserved from '../components/Rockets/RocketsReserved';
import store from '../redux/configureStore';

describe('Components test', () => {
  test('RocketCard test', () => {
    const element = (
      <Provider store={store}>
        <RocketCard
          key={1}
          id={1}
          name="Falcon 1"
          description="The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth."
          reserved={false}
          image="https://imgur.com/DaCfMsj.jpg"
        />
      </Provider>
    );
    const card = renderer.create(element);
    expect(card).toMatchSnapshot();
  });

  test('RocketsList test', () => {
    const rocketsInfo = [
      {
        id: 1, name: 'Falcon 1', description: 'Short description', reserved: false, image: 'https://imgur.com/DaCfMsj.jpg',
      },
      {
        id: 2, name: 'Falcon 2', description: 'Short description', reserved: false, image: 'https://imgur.com/DaCfMsj.jpg',
      },
    ];
    const element = (
      <Provider store={store}>
        <RocketList rockets={rocketsInfo} />
      </Provider>
    );
    const list = renderer.create(element);
    expect(list).toMatchSnapshot();
  });

  test('Reserved Rockets test', () => {
    const rocketsInfo = [
      {
        id: 1, name: 'Falcon 1', description: 'Short description', reserved: true, image: 'https://imgur.com/DaCfMsj.jpg',
      },
      {
        id: 2, name: 'Falcon 2', description: 'Short description', reserved: true, image: 'https://imgur.com/DaCfMsj.jpg',
      },
    ];
    const element = (<RocketsReserved rockets={rocketsInfo} />);
    const list = renderer.create(element);
    expect(list).toMatchSnapshot();
  });

  test('Rockets rendering test', () => {
    // this test only test that the component can be rendered
    const element = (
      <Provider store={store}>
        <Rockets />
      </Provider>
    );
    const rockets = renderer.create(element);
    expect(rockets).toMatchSnapshot();
  });
});
