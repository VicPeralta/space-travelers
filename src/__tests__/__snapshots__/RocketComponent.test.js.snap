// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components test Reserved Rockets test 1`] = `
<div
  className="reserved-container"
>
  <h2>
    My Rockets
  </h2>
  <ul
    className="rockets-reserved"
  >
    <li>
      Falcon 1
    </li>
    <li>
      Falcon 2
    </li>
  </ul>
</div>
`;

exports[`Components test RocketCard test 1`] = `
<div
  className="rocket-card"
>
  <img
    alt="rocket type"
    height={150}
    src="https://imgur.com/DaCfMsj.jpg"
    width={200}
  />
  <div
    className="info-rocket"
  >
    <h3>
      Falcon 1
    </h3>
    <p>
      The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
    </p>
    <button
      className="reserved-button"
      data-id={1}
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;

exports[`Components test Rockets rendering test 1`] = `
<h3>
  Fetching data please wait...
</h3>
`;

exports[`Components test RocketsList test 1`] = `
<div
  className="rocketContainer"
>
  <div
    className="rocket-card"
  >
    <img
      alt="rocket type"
      height={150}
      src="https://imgur.com/DaCfMsj.jpg"
      width={200}
    />
    <div
      className="info-rocket"
    >
      <h3>
        Falcon 1
      </h3>
      <p>
        Short description
      </p>
      <button
        className="reserved-button"
        data-id={1}
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
  <div
    className="rocket-card"
  >
    <img
      alt="rocket type"
      height={150}
      src="https://imgur.com/DaCfMsj.jpg"
      width={200}
    />
    <div
      className="info-rocket"
    >
      <h3>
        Falcon 2
      </h3>
      <p>
        Short description
      </p>
      <button
        className="reserved-button"
        data-id={2}
        onClick={[Function]}
        type="button"
      >
        Reserve Rocket
      </button>
    </div>
  </div>
</div>
`;
